### ISSUE TITLE
[FSDP] .modules() return original modules instead of FSDP prefixed modules

### ISSUE THREAD

**Issue by zhaojuanmao** [pytorch/pytorch#78045]
### 🚀 The feature, motivation and pitch

FSDP destructively recreated a module and changed the module path, but users still expect to get the original module path in their application, override the .module() API to return the original module with a clean path. Similar to https://github.com/pytorch/pytorch/pull/74333, which overrides named_parameters() API to return named parameters with the clean paths. 

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

**Comment by LinjianMa** [comment]
I can think of three different ways to implement this function, below are my thoughts:

1. Similar to #74333, when the state is `SUMMON_FULL_PARAMS`, remove the FSDP wraps in the module, and then return an iterator containing all modules in the unwrapped module. However, the step to remove FSDP wraps is an in-place update, so the module will not have FSDP wraps for all later calculations, which is not good.
2. Similar to 1, but first perform a deep-copy of the module, and then remove the FSDP wraps of the copied module and return it. The advantage is that this will not affect the module itself, but deep-copy will also copy all the parameters and is not very memory-efficient.
3. Since I think the only situation we want to remove the FSDP wraps in the module is when the state is `SUMMON_FULL_PARAMS`, another way we can do is to in-place remove the FSDP wraps when we enter the `_summon_full_params` contextmanager, and in-place add the FSDP wraps back when we exit the context manager. In that way we don’t need to override the `.modules()` API, since that will automatically output modules we want (when the state is `SUMMON_FULL_PARAMS` the module doesn’t contain FSDP wraps, otherwise the FSDP wraps are there). I also think in this way we also don’t need to override `named_parameters()` and `named_buffers()`.

To me 3 is a better way to implement this, please let me know if you have any suggestions :)

**Comment by zhaojuanmao** [comment]
synced up offline, will see whether we can return a list of original module references in summon_full_params context manager

**Comment by LinjianMa** [comment]
Consider a simple one layer module as follows,
```python
class OneLayerNetModule(torch.nn.Module):
    def __init__(self, D_in, H, D_out):
        super(OneLayerNetModule, self).__init__()
        self.linear1 = torch.nn.Linear(D_in, H)
```
after we construct a `fsdpmodule`,
```python
    fsdpmodel = FSDP(module, auto_wrap_policy=always_wrap_policy)
```
the output of `fsdpmodel.modules()` would output 6 modules,
```python
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): OneLayerNetModule(
      (linear1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Linear(in_features=3, out_features=2, bias=True)
        )
      )
    )
  )
)
FlattenParamsWrapper(
  (_fpw_module): OneLayerNetModule(
    (linear1): FullyShardedDataParallel(
      (_fsdp_wrapped_module): FlattenParamsWrapper(
        (_fpw_module): Linear(in_features=3, out_features=2, bias=True)
      )
    )
  )
)
OneLayerNetModule(
  (linear1): FullyShardedDataParallel(
    (_fsdp_wrapped_module): FlattenParamsWrapper(
      (_fpw_module): Linear(in_features=3, out_features=2, bias=True)
    )
  )
)
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): Linear(in_features=3, out_features=2, bias=True)
  )
)
FlattenParamsWrapper(
  (_fpw_module): Linear(in_features=3, out_features=2, bias=True)
)
Linear(in_features=3, out_features=2, bias=True)
```
Ideally, within the summon_full_params context manager, we expect the output modules to be 
```python
OneLayerNetModule(
  (linear1): Linear(in_features=3, out_features=2, bias=True)
)
Linear(in_features=3, out_features=2, bias=True)
```
To get the two modules above, we need to recursively removing the FSDP wraps in the fsdpmodule. Without deepcopy of the module, it would be an inplace update which will change the state of the module even out of the summon_full_params context manager
